### ISSUE TITLE
[Bug]: `get_ticklocs()` returns ticks outside axis limits for log scale

### ISSUE THREAD

**Issue by dstansby** [matplotlib/matplotlib#24572]
### Bug summary

`get_ticklocs()` is returning ticks outside the axis limits, when the scale is set to log.

I discovered this when writing some tests for `LogLocator` in https://github.com/matplotlib/matplotlib/pull/24436.

### Code for reproduction

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()

ax.set_xscale('log')
ax.set_xlim(1, 99)

fig.canvas.draw()
print(ax.xaxis.get_ticklocs())
# [1.e-01 1.e+00 1.e+01 1.e+02 1.e+03]
plt.show()
```

### Actual outcome

Prints `[1.e-01 1.e+00 1.e+01 1.e+02 1.e+03]`, but plots this:

![log_ticks](https://user-images.githubusercontent.com/6197628/204923640-b5f2b1cc-cdb5-4192-ab59-522d347fc0fe.png)

Note that the only ticks shown on this graph are 1 and 10.

### Expected outcome

`ax.xaxis.get_ticklocs()` returns only tick locations present on the axis.

### Matplotlib Version

3.7.0.dev656+g79807c34ad.d20221130

**Comment by dstansby** [comment]
To document where I've got with this:

- The logic in `LogLocator` for returning lots of ticks outside the range of `vmin, vmax` was added in https://github.com/matplotlib/matplotlib/pull/1686, to solve https://github.com/matplotlib/matplotlib/issues/1310. It looks like no tests were added in that PR for the original issue in https://github.com/matplotlib/matplotlib/issues/1310, and that issue seems unrelated to `LogLocator`.
- I've updated the logic in `LogLocator` to return ticks inside `vmin, vmax` in https://github.com/matplotlib/matplotlib/pull/24436. I think this is fine, and fixes this issue, but will wait and see if there are any test failures.
